File: project_euler/problem_89/roman_numeral_reduction.py
import sys

DEFAULT_FILE = "roman.txt"
# string replacement rules drawn from problem; the 5 character rules must be searched and replaced first
replacementRules = [
    ("VIIII", "IX"),
    ("IIII", "IV"),
    ("LXXXX", "XC"),
    ("XXXX", "XL"),
    ("DCCCC", "CM"),
    ("CCCC", "CD"),
]

class RomanNumeralReduction:

    def __init__(self, filePath):
        self.filePath = filePath

    #loop through rule list and replace any matching strings with their shorter equivalent
    def minimizeNumeral(self, romanNumeral):
        reducedNumeral = romanNumeral
        for long, short in replacementRules:
            reducedNumeral = reducedNumeral.replace(long, short)
        return reducedNumeral

    #loop through lines of file and apply the reduction, summing up the character savings for each line
    def doReduction(self):
        with open(self.filePath, 'r') as file:
            savings = 0
            for romanNumeral in file:
                romanNumeral.strip() #remove whitespaces
                savings += len(romanNumeral) - len(self.minimizeNumeral(romanNumeral)) #sum up character savings for each line
            return savings

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filePath = DEFAULT_FILE
    else:
        filePath = sys.argv[1]
    romanNumeral = RomanNumeralReduction(filePath)
    print(romanNumeral.doReduction())

File: project_euler/problem_89/RomanNumeralReduction.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.List;
import static java.util.Arrays.*;

public class RomanNumeralReduction
{

    private String filePath = "roman.txt";
    private static final List<List<String>> replacementRules =
        asList(
            asList("VIIII", "IX"),
            asList("IIII", "IV"),
            asList("LXXXX", "XC"),
            asList("XXXX", "XL"),
            asList("DCCCC", "CM"),
            asList("CCCC", "CD")
        );

    public RomanNumeralReduction(){}

    public RomanNumeralReduction(String filePath)
    {
        this.filePath = filePath;
    }

    public static void main(String[] args)
    {
        RomanNumeralReduction romanNumeralReduction;
        if(args.length > 0)
        {
            romanNumeralReduction = new RomanNumeralReduction(args[0]);
        }
        else
        {
            romanNumeralReduction = new RomanNumeralReduction();
        }

        try
        {
            System.out.println(romanNumeralReduction.doReduction());
        }
        catch(IOException e)
        {
            e.printStackTrace();
        }
    }

    private String minimizeNumeral(String romanNumeral)
    {
        String reducedNumeral = romanNumeral;
        for(List<String> rule : replacementRules)
        {
            reducedNumeral = reducedNumeral.replaceAll(rule.get(0), rule.get(1));
        }
        return reducedNumeral;
    }

    private int doReduction() throws IOException
    {
        int savings = 0;
        try(BufferedReader br = new BufferedReader(new FileReader(this.filePath)))
        {
            String romanNumeral = br.readLine();
            while(romanNumeral != null)
            {
                romanNumeral = romanNumeral.replaceAll(" ", "");
                savings += romanNumeral.length() - minimizeNumeral(romanNumeral).length();
                romanNumeral = br.readLine();
            }
        }

        return savings;
    }
}
